### DISCUSSION TITLE
Joint models

### DISCUSSION THREAD

**Discussion by clragon** [simolus3/drift#2752] [Q&A] [answered]
Because drifts model generation is very convenient and I have been ignoring it in favour of my own freezed classes, 
I am looking into leaning more into the model generation.

However, I cannot figure out how I could generate a model that is the result of a join.
Assume this completely fictional table setup:

```dart

class DungeonsTable extends Table {
  IntColumn get id => integer().autoIncrement()();
}

class BossRoomsTable extends Table {
  IntColumn get id =>
      integer().references(DungeonsTable, #id, onDelete: KeyAction.cascade)();
  IntColumn get boss => integer()();
}

class Boss extends Table {
  IntColumn get id => integer().autoIncrement()();
  TextColumn get name => text()();
  TextColumn get description => text()();
  TextColumn get image => text()();
}

class TreasuresRoomTable extends Table {
  IntColumn get id =>
      integer().references(DungeonsTable, #id, onDelete: KeyAction.cascade)();
  IntColumn get treasure => integer()();
}

class Treasure extends Table {
  IntColumn get id => integer().autoIncrement()();
  TextColumn get name => text()();
  TextColumn get description => text()();
  TextColumn get image => text()();
}
```

now it would be really nice if I could automatically generate a class like
`Dungeon`, `BossRoom` and `TreasureRoom` that have a structure like

```dart
class Dungeon {
  final BossRoom bossRoom;
  final TreasureRoom treasureRoom;
}

class BossRoom {
  final Boss boss;
}

class TreasureRoom {
  final Treasure treasure;
}
```

which are essentially just a bunch of joins.
I can of course manually write all this, but it would be really cool if drift would let me generate it instead.

I have briefly looked into Views on whether I can pull off something like that, but I was unable to quite understand them and I believe they might not be suited.

**Comment by simolus3** [comment, marked as answer]
You're right - views would generate flat data structures so they can't do this.

In my opinion, drift's generator should only generate row classes for tables without going into the direction of a full ORM. Resolving references in row classes and queries is not in scope for this project. I think drift is a great platform for others to build this on: Drift builders emit their results as hidden `json` files as well which others could pick up to infer relations and the runtime API can be used to build complex queries too. But I don't think this should be part of drift itself.

If you don't mind writing the queries yourself, you can use [nested results](https://drift.simonbinder.eu/docs/sql-api/drift_files/#nested-results) in drift files which would give you a similar structure.

**Reply by clragon** [reply]
> In my opinion, drift's generator should only generate row classes for tables without going into the direction of a full ORM

Thats a shame because I would like to somehow have some kind of consistency.
Either I make my models or drift makes my models.

I guess in this case its _me_ making my models, the joint ones too.

I feel like drift not generating models for joins seems somewhat incomplete; 
a joint model seems like an easy requirement to have and so I now need to bring in other generators to make those, besides drift.
But I understand why you would not want to write that hornets nest too.

> in drift files

unfortunately I am a big fan of writing everything in dart without drift files.
